File: client/audio/output_limiter.py
from __future__ import annotations

import math


class OutputLimiter:
    def __init__(
        self,
        threshold: float = 0.82,
        release: float = 0.004,
        soft_clip_drive: float = 1.25,
    ) -> None:
        self.threshold = max(0.1, min(0.99, threshold))
        self.release = max(0.0001, min(1.0, release))
        self.soft_clip_drive = max(0.1, soft_clip_drive)
        self._gain = 1.0
        self._soft_clip_normalizer = math.tanh(self.soft_clip_drive)

    def reset(self) -> None:
        self._gain = 1.0

    def process(self, samples: list[float]) -> list[float]:
        if not samples:
            return []
        processed: list[float] = []
        gain = self._gain
        for sample in samples:
            absolute_sample = abs(sample)
            target_gain = 1.0
            if absolute_sample > self.threshold and absolute_sample > 1.0e-9:
                target_gain = self.threshold / absolute_sample
            if target_gain < gain:
                gain = target_gain
            else:
                gain += (1.0 - gain) * self.release
                gain = min(1.0, gain)
            limited = sample * gain
            if abs(limited) > self.threshold:
                scaled = limited / self.threshold
                limited = (
                    math.tanh(scaled * self.soft_clip_drive) / self._soft_clip_normalizer
                ) * self.threshold
            processed.append(max(-1.0, min(1.0, limited)))
        self._gain = gain
        return processed

File: native/MaydayAudioHost/OutputLimiter.cs
namespace MaydayAudioHost;

internal sealed class OutputLimiter
{
    private readonly double _threshold;
    private readonly double _release;
    private readonly double _softClipDrive;
    private readonly double _softClipNormalizer;
    private double _gain = 1.0;

    public OutputLimiter(double threshold = 0.82, double release = 0.004, double softClipDrive = 1.25)
    {
        _threshold = Math.Clamp(threshold, 0.1, 0.99);
        _release = Math.Clamp(release, 0.0001, 1.0);
        _softClipDrive = Math.Max(0.1, softClipDrive);
        _softClipNormalizer = Math.Tanh(_softClipDrive);
    }

    public void Reset()
    {
        _gain = 1.0;
    }

    public float[] Process(float[] samples)
    {
        if (samples.Length == 0)
        {
            return samples;
        }

        var output = new float[samples.Length];
        var gain = _gain;
        for (int index = 0; index < samples.Length; index++)
        {
            var sample = samples[index];
            var absolute = Math.Abs(sample);
            double targetGain = 1.0;
            if (absolute > _threshold && absolute > 1.0e-9)
            {
                targetGain = _threshold / absolute;
            }

            if (targetGain < gain)
            {
                gain = targetGain;
            }
            else
            {
                gain += (1.0 - gain) * _release;
                gain = Math.Min(1.0, gain);
            }

            var limited = sample * gain;
            if (Math.Abs(limited) > _threshold)
            {
                var scaled = limited / _threshold;
                limited = (Math.Tanh(scaled * _softClipDrive) / _softClipNormalizer) * _threshold;
            }

            output[index] = (float)Math.Clamp(limited, -1.0, 1.0);
        }

        _gain = gain;
        return output;
    }
}
